### PR TITLE
reimplement std.traits.functionLinkage using __traits

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

**Comment by WalterBright** [comment]
Blocked by https://github.com/dlang/dmd/pull/6823

**Comment by WalterBright** [comment]
blocked by https://github.com/dlang/dmd/pull/6823

File: std/traits.d [hunk 1]
```diff
@@ -1853,21 +1853,16 @@ template isUnsafe(alias func)
...
 
 /**
-Returns the calling convention of function as a string.
+Determine the linkage attribute of the function.
+Params:
+    func = the function symbol, or the type of a function, delegate, or pointer to function
+Returns:
+    one of the strings "D", "C", "Windows", "Pascal", or "Objective-C"
 */
 template functionLinkage(func...)
```
Comment (UplinkCoder) [inline std/traits.d]: while you are at it, make func... into alias fn

File: std/traits.d [hunk 2]
```diff
@@ -1853,21 +1853,16 @@ template isUnsafe(alias func)
...
-
-    enum string functionLinkage =
-        [
-            'F': "D",
-            'U': "C",
-            'W': "Windows",
-            'V': "Pascal",
-            'R': "C++"
-        ][ mangledName!Func[0] ];
+    enum string functionLinkage = __traits(getLinkage, FunctionTypeOf!func);
```
Comment (UplinkCoder) [inline std/traits.d]: could you do a static if ?
such that we do not instanciate function-typeof if the type if a function ? FunctionTypeOf

File: std/traits.d [hunk 3]
```diff
@@ -1853,21 +1853,16 @@ template isUnsafe(alias func)
 
 
 /**
-Returns the calling convention of function as a string.
+Determine the linkage attribute of the function.
+Params:
+    func = the function symbol, or the type of a function, delegate, or pointer to function
+Returns:
+    one of the strings "D", "C", "Windows", "Pascal", or "Objective-C"
```
Comment (rainers) [inline std/traits.d]: Where is "C++" in the list possible linkage types?

**Pull request merged**